[FLINK-13550][rest][ui] Vertex Flame Graph

afedulov opened the pull request:
## What is the purpose of the change

* Adds Flame Graphs to Flink in accordance to FLIP-165 [1].


## Brief change log

  - Adds a new REST handler to `WebMonitorEnpoint`
  - Adds a tracker as a results caching layer (`ThreadInfoOperatorTracker`)
  - Adds a coordinator for sampling `ThreadInfo` of parallel operator instances and assembling the results
  - Extends `TaskExecutor` method for sampling `ThreadInfo` for a particular `Task`
  - Adds `ThreadInfoSampleService` for repeated sampling of a specific `Task` 
  - Extends `JvmUtils` to enable data collection from a specific thread
  - Adds data structures for assembling and transferring the results (`ThredInfoSample`, `TaskThreadInfoResponse`, `OperatorThreadInfoStats`, `OperatorFlameGraph`)
  - Adds UI for displaying the sampling results

## Verifying this change

This change added tests and can be verified as follows:
  - Added tests that validate that results are tracked (cached) correctly
  - Added tests that validate that samples from the parallel tasks are assembled correctly
  - Added tests sampling of individual threads is performed correctly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes, in the UI)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (/ops/debugging/flame_graphs)
  

P.S: I would like to give kudos to @dmvk for his prototyping work [2] on this feature. Although the proposed implementation significantly diverges from his prototype on the Flink side, the work done on connecting the d3-flame-graph library to the right data structure retrieved from Flink was instrumental for enabling this feature.

---
[1] https://cwiki.apache.org/confluence/display/FLINK/FLIP-165%3A+Operator%27s+Flame+Graphs
[2] https://issues.apache.org/jira/browse/FLINK-13550?focusedCommentId=17083026&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17083026

afedulov commented:
@flinkbot run azure

afedulov commented:
@flinkbot run azure

afedulov commented:
@flinkbot run azure

alpinegizmo commented on docs/content/docs/ops/debugging/flame_graphs.md:
```diff
@@ -0,0 +1,83 @@
+
+# Flame Graphs
+
+[Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are a visualization that effectively surfaces answers to the questions like:
```
```suggestion
[Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are a visualization that effectively surfaces answers to questions like:
```

alpinegizmo commented on docs/content/docs/ops/debugging/flame_graphs.md:
```diff
@@ -0,0 +1,83 @@
+
+[Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are a visualization that effectively surfaces answers to the questions like:
+- Which methods are currently consuming CPU resources?
+- How consumption by one method compares to the others?
```
```suggestion
- How does consumption by one method compare to the others?
```

alpinegizmo commented on docs/content/docs/ops/debugging/flame_graphs.md:
```diff
@@ -0,0 +1,83 @@
+
+##  Sampling process
+
+Stack traces collection happens purely within a JVM, so only method calls within the Java runtime are visible (no system calls).
```
```suggestion
The collection of stack traces is done purely within the JVM, so only method calls within the Java runtime are visible (no system calls).
```

alpinegizmo commented on docs/content/docs/ops/debugging/flame_graphs.md:
```diff
@@ -0,0 +1,83 @@
+
+{{< hint info >}}
+**Note:** 
+Stack trace samples from all threads of an operator are combined together. If a method call consumes 100%  of resources in one of the parallel tasks but none in the others, the bottleneck might be obcured by averaging out.   
```
```suggestion
Stack trace samples from all threads of an operator are combined together. If a method call consumes 100% of the resources in one of the parallel tasks but none in the others, the bottleneck might be obscured by being averaged out.   
```

alpinegizmo reviewed (commented):
Thank you for providing docs with this PR! 

I've offered a few minor suggestions.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java:
```diff
@@ -0,0 +1,323 @@
+     * @param <T> Type of the result collected from tasks.
+     * @param <V> Type of the result assembled and returned when all tasks where sampled.
+     */
+    protected abstract static class PendingStatsRequest<T, V> {
```
Move to top-level? I'm finding nested abstracts odd if you do not include nested implementations.

afedulov replied:
The idea here was that different `TaskStatsRequestCoordinators` sort of always come together as a package with the corresponding internal `PendingStatsRequest` implementations. See `TaskStatsRequestCoordinator`. In my mind such grouping underlines that <T,V> are the same thing in both classes. It made sense to me, but I do not have any hard arguments against if you really want to split them.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java:
```diff
@@ -0,0 +1,323 @@
+     * A pending task stats request, which collects samples from individual tasks and completes the
+     * response future upon gathering all of of them.
+     *
+     * <p>Has to be accessed in lock scope.
```
Might be easier to add `@NotThreadSafe`.

afedulov replied:
added the annotation.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,188 @@
+     * @return A future of the completed thread info stats.
+     */
+    public CompletableFuture<OperatorThreadInfoStats> triggerThreadInfoRequest(
+            List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
```
This parameter type is inducing nightmares ;) Maybe add a small record class?

zentol replied:
This code would be easier to test if it wouldn't specifically require a TaskExecutorGateway; we could move the newly added method into a separate interface that the TM extends.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,188 @@
+                                    executorGateway.requestThreadInfoSamples(
+                                            taskExecutionAttemptId, requestParams, timeout));
+
+            threadInfo.handleAsync(
```
whenCompleteAsync?

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleableTask.java:
```diff
@@ -0,0 +1,30 @@
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+
+/** Task interface used by {@link ThreadInfoSampleService} for thread info tracking. */
+interface ThreadInfoSampleableTask {
```
Maybe just `SampleableTask`?

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -74,6 +74,8 @@
 import org.apache.flink.runtime.management.JMXService;
 import org.apache.flink.runtime.memory.MemoryManager;
 import org.apache.flink.runtime.messages.Acknowledge;
+import org.apache.flink.runtime.messages.TaskThreadInfoResponse;
```
For commit message, I'd reformulate it in a positive way: you need it for test mocks.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleService.java:
```diff
@@ -0,0 +1,107 @@
+/*
```
For such helper commits, it's always good to state when they are used. The `ThreadInfoSampleService` is later injected into ... to ....

AHeise commented on flink-runtime-web/web-dashboard/package.json:
```diff
@@ -22,6 +22,8 @@
     "@antv/g2": "^3.4.10",
     "core-js": "^2.5.4",
     "d3": "^5.9.1",
+    "d3-flame-graph": "^4.0.6",
```
I have no clue what's going on here ;) But it looks good.

One thing that I don't understand at all - why do you update package-lock in this commit and then in the next commit with a huge diff?

afedulov replied:
There is `package.json` which is for user-defined settings and `package-lock.json`, which is auto-generated upon commit `npm build`.

AHeise replied:
Okay thanks for explaining. Not sure why we commit something autogenerated but afaik you didn't add it.

afedulov replied:
This seems to be the accepted practice for node: https://stackoverflow.com/a/44210813/336152 . I do not understand why the whole file get's changed completely though in this case - I believe I've seen smaller patches for it in the commit history, which would indicate that it was in fact edited manually ¯\_(ツ)_/¯.

AHeise commented on flink-runtime-web/web-dashboard/package.json:
```diff
@@ -43,11 +45,11 @@
     "husky": "^1.3.1",
     "jasmine-core": "~2.99.1",
     "jasmine-spec-reporter": "~4.2.1",
+    "karma": "~4.0.0",
```
Reordering probably makes it much harder to track what packages you changed here.

afedulov replied:
Thanks for spotting. Seems that my editor applied different ordering rules for strings of variable length than what was in the file originally.

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerTest.java:
```diff
@@ -166,6 +166,34 @@ public void testRequestTaskManagerInfo() throws Exception {
         assertEquals(0, taskManagerInfo.getNumberAvailableSlots());
     }
 
+    /**
```
Tests should be part of the respective main commits. You want each commit to stand on its own feet (that's not always strictly possible).

AHeise reviewed (commented):
This new feature will be awesome. Lots of users will appreciate your work, so thank you very much.

There are a couple of things that should be address before a merge. Before going into details, I must state that I'm not familiar with most parts of the code and design pattern, so take everything with a grain of salt. It's best if @zentol does another pass - he would be a much better reviewer.

I think the overall approach and architecture is very plausible. I found a couple of smaller things:
- Commit structure needs fine-tuning: 1-2 commits are too small but most importantly 2 commits are oddly cut: the data structure commit and test commit seem to be cross-cutting concerns and should be split and merged into the respective commits.
- All commit messages could use a few sentences on the purpose to guide the reviewer. That's especially important after merge when someone looks at the individual commits.
- Some commits are not self-contained and reference classes added later. Commits should always compile and tests should be green.
- I'm not convinced that we need to update the `package-lock` without knowing much about it.
- I'm not 100% sure if we need that cache in the flamegraph handler, especially if the user has to set cache cleanup interval. This might be the only questionable design choice in this PR from my point of view. Web ui could simply poll asynchronously (which it probably does) and is getting uncached result as soon as it arrives. But I might have misunderstood the purpose.
- Please check my comments on config options. There may be alternative ways on how to present it to the user.
- Please check if the components of all commits is correct. It should be mostly [rest].

I have not looked at the docs, generated docs, ui, and test commits. I'd look at the test once it's restructured, so it's easier for me to connect the test with the code that was just introduced into a commit. I'll probably review docs after feature freeze for time reasons. Please find someone that can review UI.

afedulov commented on flink-core/src/main/java/org/apache/flink/configuration/WebOptions.java:
```diff
@@ -156,6 +156,48 @@
                     .withDescription("This config option is no longer used");
 
+    /** Time, in milliseconds, after which cached stats are cleaned up if not accessed. */
+    public static final ConfigOption<Integer> FLAMEGRAPH_CLEANUP_INTERVAL =
```
Do you mean switching to `java.time.Duration` ?

AHeise replied:
Yes. That makes it unnecessary to specify the time unit in docs.

afedulov replied:
Not sure I get it. In the config file a user would still specify just a number, right? And we need to document in which units we are going to interpret this number. Like, say `ExecutionConfig.BUFFER_TIMEOUT`. What am I missing?

afedulov replied:
Got it, thanks. Switched to Duration.

zentol replied:
Durations where possible. Flinks Time class is effectively deprecated in favor of standard JDK classes.

afedulov replied:
Moved to using Duration.

afedulov commented on flink-core/src/main/java/org/apache/flink/configuration/WebOptions.java:
```diff
@@ -156,6 +156,48 @@
                     .withDeprecatedKeys("jobmanager.web.backpressure.delay-between-samples")
                     .withDescription("This config option is no longer used");
 
+    /** Time, in milliseconds, after which cached stats are cleaned up if not accessed. */
```
I understand the intent, but I think it might add some potential for confusion with respect to how the sampling works. Mainly because window semantics in Flink is based on time, whereas here it is about gathering a certain number of data points. Not sure it is fits.

AHeise replied:
You also mentioned offline that this is similar to how backpressure sampling works. So if a user already knows that, it's better to keep it in this way.

afedulov replied:
Yes, this approach is largely inspired by the previously existing backpressure monitor which was also based on stack traces sampling. A couple of pointers:
https://github.com/apache/flink/blob/c685251ce3b39362e45b07a12aa0cc6e7f665021/flin[…]re/src/main/java/org/apache/flink/configuration/WebOptions.java
https://github.com/apache/flink/blob/7e4694b8198664300b13b5304c62271b192d4512/flin[…]pache/flink/runtime/webmonitor/StackTraceSampleCoordinator.java
https://github.com/apache/flink/blob/7e4694b8198664300b13b5304c62271b192d4512/flin[…]g/apache/flink/runtime/webmonitor/BackPressureStatsTracker.java

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexFlameGraphHandler.java:
```diff
@@ -0,0 +1,121 @@
+
+        switch (flameGraphType) {
+            case FULL:
+                operatorFlameGraph =
```
I did not want to introduce a dependency of `WebOptions[flink-core]->OperatorFlameGraphFactory[flink-runtime]`. `OperatorFlameGraphFactory` does not also seem to belong into [flink-core]. Do you think it is worth it?

AHeise replied:
Nope, it look like it was in the same package. So keep as is.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,188 @@
+                runningSubtasksIds.add(execution.getAttemptId());
+            } else {
+                return FutureUtils.completedExceptionally(
+                        new IllegalStateException(
```
Will failing so hard not creating quite a bit of noise?

afedulov replied:
Probabaly... That's what the old backpressure sampler was doing https://github.com/apache/flink/blob/7e4694b8198664300b13b5304c62271b192d4512/flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/StackTraceSampleCoordinator.java#L129

What would be the best practice here in your opinion?

AHeise replied:
Tracing on caller side now. Semantic change: Flame graphs are not aborted for tasks with finished subtasks - I think that this is useful for debugging strugglers.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,188 @@
+        // triggering can still fail.
+        for (Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>
+                executionsWithGateway : subtasksWithGateways) {
+            AccessExecution execution = executionsWithGateway.f0.getCurrentExecutionAttempt();
```
IF all we need are the attemptIds then we should only require that to be passed in.

afedulov replied:
The motivation was to pin the task which is failing in case of errors:
https://github.com/apache/flink/pull/15054/files/de5180a21e4f51384895bee7e2527677aa8092c1#diff-803556ada30870d9cec571303d2d7d89a63f9980c26a79020a15ab239e2a837eR95
So it is somewhat related to the question raised in the previous comment about error handling.

AHeise replied:
Filtering state before passing the `executionsWithGateway` now - that allows to only pass `AttemptID`.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/ThreadInfoSample.java:
```diff
@@ -0,0 +1,283 @@
+ */
+public class ThreadInfoSample implements Serializable {
+
+    private final String threadName;
+    private final long threadId;
+    private final long blockedTime;
+    private final long blockedCount;
+    private final long waitedTime;
+    private final long waitedCount;
+    private final String lockName;
+    private final long lockOwnerId;
+    private final String lockOwnerName;
+    private final boolean inNative;
+    private final boolean suspended;
+    private final Thread.State threadState;
+    private final StackTraceElement[] stackTrace;
```
Do we really need all these things?

afedulov replied:
Not for the Flame Graphs, but I was thinking that they maybe could be helpful if we further extend our debugging toolkit. Remove unused for now?

afedulov replied:
Done.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/OperatorFlameGraphFactory.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private static OperatorFlameGraph.Node buildFlameGraph(NodeBuilder builder) {
+        final List<OperatorFlameGraph.Node> children = new ArrayList<>();
```
Looking at the graphs that I produced with my fraud detection sample application it seems that the stack depth is around 20 typically. How about doubling it to account for some sophisticated user code calls? 40 would be my best guess.

AHeise replied:
You are thinking to complicated. The size `=children.values()`.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerGateway.java:
```diff
@@ -266,4 +267,12 @@ void heartbeatFromTaskManager(
+     * @param taskManagerId identifying the {@link TaskExecutor}.
+     * @return Future containing the task executor gateway.
+     */
+    CompletableFuture<TaskExecutorGateway> requestTaskExecutorGateway(ResourceID taskManagerId);
```
We generally don't expose other component gateways, because it make it unnecessarily difficult to mock the RMGateway in tests.

afedulov replied:
Flame Graphs component needs to be able to send requests to the `TaskExecutors` that are responsible for running the subtasks of the vertex selected in the UI. What would be the alternative?
P.S.: it has been an explicit request from Till to avoid adding non-core functionality to the `JobManagerSharedServices`, the way the old backpressure sampler was implemented, and rather rely on the `ArchivedExecutionGraph` which is already available to the `WebMonitorEndpoint`.

AHeise replied:
Now returning only the extracted `TaskExecutorThreadInfoGateway`.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexFlameGraphHeaders.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public String getDescription() {
+        return "Returns flame graph information for a job, and may initiate flame graph sampling if necessary.";
```
```suggestion
        return "Returns flame graph information for a vertex, and may initiate flame graph sampling if necessary.";
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexFlameGraphInfo.java:
```diff
@@ -0,0 +1,60 @@
+
+    @JsonCreator
+    public JobVertexFlameGraphInfo(
+            @JsonProperty(FIELD_NAME_END_TIMESTAMP) Long endTimestamp,
```
```suggestion
            @JsonProperty(FIELD_NAME_END_TIMESTAMP) long endTimestamp,
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexFlameGraphInfo.java:
```diff
@@ -0,0 +1,60 @@
+        this.root = root;
+    }
+
+    public Long getEndTimestamp() {
```
```suggestion
    publiclong getEndTimestamp() {
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexMessageParameters.java:
```diff
@@ -26,6 +26,8 @@
 public class JobVertexMessageParameters extends JobMessageParameters {
 
     public final JobVertexIdPathParameter jobVertexIdPathParameter = new JobVertexIdPathParameter();
+    public final FlameGraphTypeQueryParameter flameGraphTypeQueryParameter =
```
Well this is pretty jank. Why is there not a separate parameters class extending this one?

afedulov replied:
My bad! Support for different Flame Graphs type was the last minute edition and I misunderstood how those `JobVertexMessageParameters` are intended to be used. Fixing it.

AHeise replied:
Already solved.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -497,6 +505,32 @@ private void stopTaskExecutorServices() throws Exception {
+    @Override
+    public CompletableFuture<TaskThreadInfoResponse> requestThreadInfoSamples(
+            final ExecutionAttemptID taskExecutionAttemptId,
+            ThreadInfoSamplesRequest requestParams,
```
```suggestion
            final ThreadInfoSamplesRequest requestParams,
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleService.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.flink.util.Preconditions.checkNotNull;
```
inconsistent usage of null checks

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleService.java:
```diff
@@ -0,0 +1,107 @@
+                requestParams.getDelayBetweenSamples(),
+                requestParams.getMaxStackTraceDepth(),
+                new ArrayList<>(requestParams.getNumSamples()),
+                new CompletableFuture<>());
```
seems a bit odd to have this as a parameter

afedulov replied:
This method is called recursively, hundreds of times per subtask, so initing it within the method body seems wasteful. What would you propose?

AHeise replied:
I removed the return type and made it apparent that this is an `out` parameter.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/OperatorStatsTracker.java:
```diff
@@ -0,0 +1,56 @@
+     * @param vertex Operator to get the stats for.
+     * @return Statistics for an operator
+     */
+    Optional<T> getOperatorStats(AccessExecutionJobVertex vertex);
```
Javadocs are missing details on when the optional is empty

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/OperatorStatsTracker.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Cleans up the operator stats cache if it contains timed out entries.
+     *
+     * <p>The Guava cache only evicts as maintenance during normal operations. If this handler is
+     * inactive, it will never be cleaned.
```
Good catch, it bubbled up during the interface extraction.

zentol reviewed (commented):
Did a quick skim, will take a deeper look later on.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,403 @@
+        this.createStatsFn = checkNotNull(createStatsFn, "Create stats function");
+        this.executor = checkNotNull(executor, "Scheduled executor");
+
+        checkArgument(cleanUpInterval >= 0, "Clean up interval");
```
Very good question. Guava will effectively disable caching if the clean-up interval is 0 [1]:
```
When duration is zero, this method hands off to maximumSize(0), ignoring any otherwise-specificed maximum size or weight. This can be useful in testing, or to disable caching temporarily without a code change.
``` https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#expireAfterAccess(long,%20java.util.concurrent.TimeUnit)

This could be useful, if we want to give users more control from the UI with regards to how often to force resampling of the results. As this functionality is currently not supported by the UI, it probably makes sense to only allow values >0 for now. What do you think?

afedulov replied:
And probably also the same for the other parameters that currently accept 0s.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,403 @@
+        synchronized (lock) {
+            final T stats = operatorStatsCache.getIfPresent(vertex);
+            if (stats == null
+                    || System.currentTimeMillis() >= stats.getEndTime() + statsRefreshInterval) {
```
Yes. I guess we should at least mention this in the docs?

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,403 @@
+     * Completed stats. Important: Job vertex IDs need to be scoped by job ID, because they are
+     * potentially constant across runs messing up the cached data.
+     */
+    private final Cache<AccessExecutionJobVertex, T> operatorStatsCache;
```
Discussed here: 
https://github.com/apache/flink/pull/15054#discussion_r602081429

Generally speaking, the purpose is to make debugging efforts somewhat less Heisenberg-y. It makes sure that if a bunch of people gets alerted about an issue with a Flink job and all start sampling stack traces in parallel, that this does not significantly further degrade the performance. Similar argument could be made about opening multiple UI windows even from the same machine. Sampling is not super expensive, but is it also does not come for free.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,403 @@
+                                            delayBetweenSamples,
+                                            maxThreadInfoDepth));
+
+            sample.handleAsync(new ThreadInfoSampleCompletionCallback(vertex), executor);
```
Good point! Switched to `BiConsumer`.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java:
```diff
@@ -0,0 +1,323 @@
+ * @param <V> Type of the combined response.
+ */
+public class TaskStatsRequestCoordinator<T, V> {
+    protected final Logger log = LoggerFactory.getLogger(getClass());
```
I needed logs both in the parent class methods and in the implementations, hence the dynamic logger initialization via `getClass()`. `getClass()` cannot be referenced from the static context.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/ThreadInfoSample.java:
```diff
@@ -0,0 +1,283 @@
+import java.lang.management.ThreadMXBean;
+
+/**
+ * A serializable wrapper around {@link java.lang.management.ThreadInfo} that excludes {@link
```
OK, I propose then to just remove them. There is already a link to `java.lang.management.ThreadInfo` at the class level, I does not seem to make much sense to link every individual methos. What do you think?

afedulov replied:
Removed the comments and limited the scope to only those fields that are currently utilized.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java:
```diff
@@ -0,0 +1,323 @@
+public class TaskStatsRequestCoordinator<T, V> {
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    protected final int numGhostSampleIds = 10;
```
This got inherited from the backpressure sampling coordinator. It is used to distinguish between replies for requests that where previously issued but cancelled vs replies who's origin is unknown.

afedulov replied:
Turned into a constant.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java:
```diff
@@ -0,0 +1,323 @@
+     * @param requestTimeout Request time out of a triggered task stats request.
+     */
+    public TaskStatsRequestCoordinator(Executor executor, long requestTimeout) {
+        checkArgument(requestTimeout >= 0L, "The request timeout must be non-negative.");
```
As described [here](https://github.com/apache/flink/pull/15054/files#diff-803556ada30870d9cec571303d2d7d89a63f9980c26a79020a15ab239e2a837eR49), it is added on top of the expected duration for sampling, so techinally it can also be 0: 
https://github.com/apache/flink/pull/15054/files#diff-803556ada30870d9cec571303d2d7d89a63f9980c26a79020a15ab239e2a837eR117
As this class is a generic abstraction that is supposed to be reused whenever sampling of any other type of statistics would be required, it seems that it would be allowing such cases. Not sure. What do you think?

afedulov commented:
@AHeise Thanks a lot for the review! I addressed all of your comments apart from those that relate to the structure of commits. There is already a lot going on in this review so I'd like to keep the current structure for now and change it when the main open points are resolved. I also kept all of your proposed changes in a separate commit for the moment.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexFlameGraphHandler.java:
```diff
@@ -0,0 +1,121 @@
+            GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+            Time timeout,
+            Map<String, String> responseHeaders,
+            MessageHeaders<EmptyRequestBody, JobVertexFlameGraphInfo, JobVertexMessageParameters>
```
So remove it as the constructor field and make it a private variable?

AHeise replied:
Already fixed.

afedulov commented:
//TODO: remove unnecessary format changes in `flink-runtime-web/web-dashboard/tsconfig.json 
Viewed`

afedulov commented:
@zentol regarding https://github.com/apache/flink/pull/15054#discussion_r602405496

> It seems a bit questionable. The licensing story around the JDK is a bit complicated, so let's be as pessimistic as possible.

Is this regarding the javadocs or does it go further than that?

afedulov commented on docs/layouts/shortcodes/generated/rest_v1_dispatcher.html:
```diff
@@ -3636,6 +3656,16 @@
+    <tr>
+      <td colspan="2">
+        <ul>
+<li><code>type</code> (optional): String value that specifies the Flame Graph type. Supported options are: "[FULL, ON_CPU, OFF_CPU]".</li>
```
Yeah, that is the result of the "jank" spotted here: https://github.com/apache/flink/pull/15054#discussion_r602421927

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/OperatorFlameGraph.java:
```diff
@@ -0,0 +1,96 @@
+    public static class Node {
+
+        private final String name;
+        private final int value;
```
This one would be a bit tricky, because the UI library enforces this format:
https://github.com/spiermar/d3-flame-graph#input-format
I could not find a straightforward way to change that without patching the library.

zentol replied:
Can we not use jackson annotations for that?

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/OperatorFlameGraphFactory.java:
```diff
@@ -0,0 +1,132 @@
+
+        private final String name;
+
+        private int value = 0;
```
see above

afedulov commented:
//TODO: remove unnecessary reformatting in `tsconfig.json`

afedulov commented:
@zentol regarding https://github.com/apache/flink/pull/15054#discussion_r602410226

> This code would be easier to test if it wouldn't specifically require a TaskExecutorGateway; we could move the newly added method into a separate interface that the TM extends.
Which class/interface do you mean by "TM"?

afedulov commented:
@zentol thanks a lot for your review! I addressed your comments in a separate commit.

afedulov commented:
@zentol thanks a lot for your review! I addressed your comments in a separate commit.

zentol commented:
> Is [the licensing issue] regarding the javadocs or does it go further than that?

I think it goes further than that. We can't copy code from the JDK because it is GPLv2 code.

> Which class/interface do you mean by "TM"?

TaskExecutorGateway

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/ThreadInfoSample.java:
```diff
@@ -0,0 +1,285 @@
+
+package org.apache.flink.runtime.messages;
+
+import com.sun.istack.Nullable;
```
replace with another annotation

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/ThreadInfoSample.java:
```diff
@@ -0,0 +1,285 @@
+     * @param threadInfo {@link ThreadInfo} where the data will be copied from.
+     * @return new {@link ThreadInfoSample}
+     */
+    public static @Nullable ThreadInfoSample from(@Nullable ThreadInfo threadInfo) {
```
This could already return an Optional

afedulov replied:
Changed.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexFlameGraphInfo.java:
```diff
@@ -0,0 +1,60 @@
+        return new JobVertexFlameGraphInfo(-1, null);
+    }
+
+    private static final String FIELD_NAME_END_TIMESTAMP = "end-timestamp";
```
```suggestion
    private static final String FIELD_NAME_END_TIMESTAMP = "endTimestamp";
```
We are trying to phase out dashes

afedulov replied:
This field it only supposed to be used in JSON and such approach seems to be an established practice:
https://github.com/apache/flink/blob/dcb9f83d4ad3710bb9e34e3de3bb82a146ff2050/flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexBackPressureInfo.java#L42

zentol replied:
As I said, camelCase is what we use for new fields. Dashes are basically just a relic from very old handlers (which also handled things inconsistently).
To reduce the amount of boilerplate we'll need once we migrate all fields please do not use dashes.

afedulov replied:
Changed as requested.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoSamplesRequest.java:
```diff
@@ -0,0 +1,91 @@
+            int requestId,
+            @Nonnegative int numSamples,
+            Time delayBetweenSamples,
+            int maxStackTraceDepth) {
```
Is this allowed to be negative? `ThreadMXBean#getThreadInfo(long, int)` seems to reject negative depths at least

afedulov replied:
Added `@Nonnegative` + check.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/OperatorFlameGraph.java:
```diff
@@ -0,0 +1,96 @@
+public class OperatorFlameGraph implements Serializable, Statistics {
+
+    /** Graph node. */
+    public static class Node {
```
Inner classes are usually placed at at the bottom

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/OperatorFlameGraphFactory.java:
```diff
@@ -0,0 +1,129 @@
+            return child;
+        }
+
+        void increment() {
```
lets rename this to something more descriptive, like incrementHitCount

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,407 @@
+    }
+
+    /** Callback on completed thread info sample. */
+    class ThreadInfoSampleCompletionCallback
```
why is this package-private?

afedulov replied:
Mistakenly inherited from the old backpressure tracker. Moved to private.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/Statistics.java:
```diff
@@ -0,0 +1,30 @@
+package org.apache.flink.runtime.webmonitor.stats;
+
+/** Represents one or more statistics samples. */
+public interface Statistics {
```
I'm not sure if we really need this interface, and the generic parameter for the ThreadInfoOperatorTracker for that matter.

afedulov replied:
The idea was mostly to surface the common pattern of sampling individual Operator subtasks and assembling the results (`OperatorStatsTracker`). I think if we continue to develop our debugging toolkit, such need will, most probably, arise again in the future.

AHeise replied:
I left as is.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,407 @@
+    public static class Builder<T extends Statistics> {
+
+        private final GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever;
+        private final Function<OperatorThreadInfoStats, T> createStatsFn;
```
unless I'm missing something this is always `Function.Identity()`.

afedulov replied:
Well, for the case of assembling the Flame Graph yes, but the idea was to be able to pass in whatever aggregation that might be needed (averaging/median etc.) for other individual parameters that could be contained in the returned samples.

AHeise replied:
I left as is.

afedulov commented:
> I think it goes further than that. We can't copy code from the JDK because it is GPLv2 code.
OK, I am not sure what to do here. It is essentially just a POJO, given the naming convention it is not really unique code.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -216,6 +227,49 @@ public WebMonitorEndpoint(
+        this.threadInfoOperatorTracker = initializeThreadInfoTracker(executor);
+    }
+
+    private ThreadInfoOperatorTracker<OperatorThreadInfoStats> initializeThreadInfoTracker(
```
maintenance-wise it would be good for this method to be static so we don't have to worry about accessing uninitialized fields if we re-order stuff (or rather to make this dependency more explicit).

afedulov replied:
As discussed in another thread, I've moved the field that get's initialized by this method into `initializeHandlers`, so it has no "side-effects". Do you think static is still necessary given the changes?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -179,6 +189,7 @@
     private final LeaderElectionService leaderElectionService;
 
     private final FatalErrorHandler fatalErrorHandler;
+    private final ThreadInfoOperatorTracker<OperatorThreadInfoStats> threadInfoOperatorTracker;
```
I think we could create this within initializeHandlers(); there isn't really a need to have this here.

afedulov replied:
moved as requested.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -216,6 +227,49 @@ public WebMonitorEndpoint(
+                                        WebOptions.FLAMEGRAPH_STACK_TRACE_DEPTH))
+                        .build();
+
+        executor.scheduleWithFixedDelay(
```
shouldn't the tracker care of care of initiating this cleanup?

afedulov replied:
I understand this as:

1. pass in the `executor` as it is done now
2. schedule the cleanup in the constructor of the tracker.

Right?

zentol replied:
yes

AHeise replied:
Done.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,187 @@
+            CompletableFuture<TaskThreadInfoResponse> threadInfo =
+                    executorGatewayFuture.thenCompose(
+                            executorGateway ->
+                                    executorGateway.requestThreadInfoSamples(
```
this will send an RPC for every attempt. To reduce the number of issues RPC calls we could batch requests for executions running in the same task executor.

afedulov replied:
Good point. Unfortunately I won't be able to address this optimization before the feature freeze because I am currently on vacation. How about we add it as an "improvement" ticket.

AHeise replied:
👍  for follow-up work.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java:
```diff
@@ -0,0 +1,407 @@
+    public Optional<T> getOperatorStats(AccessExecutionJobVertex vertex) {
+        synchronized (lock) {
+            final T stats = operatorStatsCache.getIfPresent(vertex);
+            if (stats == null
```
Where do we prevent multiple requests being initiated when the first one has not completed yet? Asking since the check is based on the operatorStatsCache that is only populated once the response from the TaskExecutor is received.

afedulov replied:
https://github.com/apache/flink/pull/15054/files/e7f59e2e2811c0718a453572e90a4ad2a900ff03#diff-c9e5dc5e2c21890cb66f2592f45b3b23462fd28f330a86a1aec824576fb60039R301  
(`ThreadInfoOperatorTracker` line 301)

zentol reviewed (changes requested):
Who has reviewed the UI portion?

zentol commented:
@vthinkxie Would you have time to review the UI changes?

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/OperatorStatsTracker.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @param <T> Type of statistics to track
+ */
+public interface OperatorStatsTracker<T extends Statistics> {
```
I actually think it is correct. We do sample individual tasks, but the tracker is responsible for assembling a combined view from them which represent an Operator (see `ThreadInfoOperatorTracker`). The main method of this interface returns such combined statistics representation for the vertex(operator):
```
Optional<T> getOperatorStats(AccessExecutionJobVertex vertex);
```

zentol replied:
Vertex != Operator; a vertex contains one or more operators.

If you'd renamed Operator with Vertex in the names then things would be fine.

afedulov replied:
Agreed. How do you think we should call a `OperatorThreadInfoStats`? `(Sub)TasksThreadInfoStats` ?

afedulov replied:
Renamed Operator -> JobVertex everywhere.

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -310,6 +315,9 @@ public TaskExecutor(
                 createResourceManagerHeartbeatManager(heartbeatServices, resourceId);
+
+        this.threadInfoSampleService =
+                new ThreadInfoSampleService(rpcService.getScheduledExecutor());
```
Which type of executor should if be in terms of pool size and where should it be initialized?

AHeise replied:
Creating a new thread pool (of size 1).

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -497,6 +505,32 @@ private void stopTaskExecutorServices() throws Exception {
+                    new IllegalStateException(
+                            String.format(
+                                    "Cannot sample task %s. "
+                                            + "Task is not known to the task manager.",
```
Do I get it correctly that this comment is primarily related to how to best handle this exception when it bubbles up to the `OperatorTracker` rather then to the fact that it is thrown here?

AHeise replied:
Left it at debug. But I have also added a check to the `JobVertexFlameGraphHandler` to short-cut to an empty result for vertexes in terminal state.

afedulov commented:
Regarding https://github.com/apache/flink/pull/15054#discussion_r602410226
> >Which class/interface do you mean by "TM"?
> TaskExecutorGateway
This seems to be a "new" approach that was not previously used for the `TaskExecutorGateway`, also for similarly-scoped methods like `requestThreadDump`. Could you expand on the idea a bit? Should it be a single method interface just for the thread info collection purposes or do you want to use it to potentially group things. What do you think would be the appropriate name?

afedulov commented:
@zentol question: by the fact that timeouts are always passed, but are never explicitly addressed in the method calls, I assume that they are handled by the underlying RPC machinery automatically? I switched the timeout type from Time to Duration - will it break the proper handling?

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -497,6 +505,32 @@ private void stopTaskExecutorServices() throws Exception {
+        return stackTracesFuture.thenApply(
+                stackTraces ->
+                        new TaskThreadInfoResponse(
+                                requestParams.getRequestId(), taskExecutionAttemptId, stackTraces));
```
This internal logic is "inherited" from the backpressure sampler that was in use for a while ([sampleId](https://github.com/apache/flink/blob/7e4694b8198664300b13b5304c62271b192d4512/flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/StackTraceSampleCoordinator.java#L161)). I agree with you that it can probably be done better, but this bit is quite at the core of the whole sampling process and I am afraid that I won't be able to address it properly before the 12th of April. Because this parameter is wrapped into `ThreadInfoSamplesRequest`, getting rid of it won't need changes of the public interfaces. Due to the time constrains I would ask to go for the "trusted" solution for now and assign a ticket to me for the improvements.

AHeise replied:
Removed request id entirely,

afedulov commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleService.java:
```diff
@@ -0,0 +1,107 @@
+            resultFuture.complete(currentTraces);
+            return resultFuture;
+        } else {
+            throw new IllegalStateException(
```
Switched from throwing an exception to completing the future exceptionally.
I do not know what is the best course of action here regarding the topic of TaskManager/TaskExecutor threads.

zentol replied:
The solution for the threading issue is fairly simple; have `requestThreadInfoSamples(SampleableTask, ThreadInfoSamplesRequest`) schedule the call to

```
resultFuture = new CompletableFuture<>();
executor.schedule(() ->
requestThreadInfoSamples(
                task,
                requestParams.getNumSamples(),
                requestParams.getDelayBetweenSamples(),
                requestParams.getMaxStackTraceDepth(),
                new ArrayList<>(requestParams.getNumSamples()),
                resultFuture));
return future;
```

AHeise replied:
Added it.

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/share/customize/flame-graph/flame-graph.component.html:
```diff
@@ -0,0 +1,23 @@
+<div class="background">
+  <div #flameGraphContainer>
+  </div>
+  {{ draw() }}
```
draw shouldn't be called inside the template, you can move it inside the `ngAfterViewInit`

AHeise replied:
Moved into component.

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/share/customize/flame-graph/flame-graph.component.ts:
```diff
@@ -0,0 +1,69 @@
+import { format } from 'd3-format';
+import { flamegraph, offCpuColorMapper } from 'd3-flame-graph';
+
+import * as _d3 from 'd3';
```
there is an angular native flamegraph here https://github.com/mgechev/ngx-flamegraph, not sure if it can meet your requirements, but you can have a try.

AHeise replied:
I'd suggest to do a follow-up PR for that.

zentol commented:
> @zentol question: by the fact that timeouts are always passed, but are never explicitly addressed in the method calls, I assume that they are handled by the underlying RPC machinery automatically? I switched the timeout type from Time to Duration - will it break the proper handling?

I'm not entirely sure but would expect that RPC timeouts must be based on Flinks Time class.

> This seems to be a "new" approach that was not previously used for the TaskExecutorGateway, also for similarly-scoped methods like requestThreadDump. Could you expand on the idea a bit? Should it be a single method interface just for the thread info collection purposes or do you want to use it to potentially group things. What do you think would be the appropriate name?

Yes, we failed to do this in the past, and paid the price for it later on :(
A SAM should suffice, for the name we could maybe use `ThreadDumpProvider`?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutorGateway.java:
```diff
@@ -266,4 +269,17 @@ void heartbeatFromJobManager(
+    CompletableFuture<TaskThreadInfoResponse> requestThreadInfoSamples(
+            ExecutionAttemptID taskExecutionAttemptId,
+            ThreadInfoSamplesRequest requestParams,
+            Duration timeout);
```
```suggestion
            @RpcTimeout Time timeout);
```

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java:
```diff
@@ -0,0 +1,187 @@
+                    (TaskThreadInfoResponse threadInfoSamplesResponse, Throwable throwable) -> {
+                        if (threadInfoSamplesResponse != null) {
+                            handleSuccessfulResponse(
+                                    threadInfoSamplesResponse.getRequestId(),
+                                    threadInfoSamplesResponse.getExecutionAttemptID(),
```
Removed request id entirely, `requestParams` has no `getExecutionAttemptID` though.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:
```diff
@@ -166,4 +168,69 @@
                                     + "increasing will allocate more time for the REST server's processing.");
+
+    /**
+     * "Time after which cached stats are cleaned up if not accessed. It can be specified using
+     * notation: "100 s", "10 m".
```
copy&paste error

AHeise commented:
AZP green. I removed one refactor commit since that broke UI. I verified manually that's working and building locally.

zzfukai commented:
Hi, I'm trying to navigate the flamegraph with the fresh commit, the "flamegraph" tab can show up, while the sever side responses `Unable to load requested file /jobs/b99a897dda4b2ab4bdf2629ea2ce300e/vertices/47d89856a1cf553f16e7063d953b7d42/flamegraph.` Any guidance on this?

AHeise commented:
We changed the feature to opt-in in the last minute and haven't updated docs yet :/. You need to set `rest.flamegraph.enabled: true` in your config to enable this feature.

zzfukai commented:
Thank you Arvid! It works.

AHeise commented:
Thanks for reporting, I'm sure @afedulov will pick it up once he is back from vacation.
